src/qibo_cloud_backends/braket_translation.py

```diff
@@ -0,0 +1,196 @@
+from functools import singledispatch
+
+from braket.circuits import Circuit as BraketCircuit
+from braket.circuits import Instruction
+from braket.circuits import gates as braket_gates
+from qibo import Circuit as QiboCircuit
+from qibo import gates as qibo_gates
+
+
+def to_braket(qibo_circuit: QiboCircuit, verbatim_circuit: bool) -> BraketCircuit:
```

Comment:
Are you sure that we need all this? Looking at the AWS [docs](https://docs.aws.amazon.com/braket/latest/developerguide/braket-openqasm-create-submit-task.html), it seems that they support direct execution of qasm code. Why not just doing:
``` python
program = Program(source=qibo_circuit.to_qasm())
task = self.device.run(
    program,
    shots=nshots,
)
```

Reply:
We did initially use OpenQASM to do the translation! But we ran into too many issues with the strings and translation with the naming. Different software providers like to use their own naming styles. For example, sqrt(x) gate is `sx` in Qibo and `SXGate` in Qiskit. These two are fairly similar. However, Braket uses `v` as the sqrt(x) gate, which is not so straightforward. 

The string produced by Qibo has strings like `includes qe1lib.inc` which are not recognized by Braket, among others. 

Our colleague from AWS suggested using a direct translation of gates using `braket_translation` to bypass reliance on OpenQASM. He mentioned that Qiskit uses the same method to translate to Braket. Any addition of gates can then be included in `braket_translation`. This way, we could avoid cluttering the `braket_client` code since the Braket circuit is generated by calling `to_braket(circuit_qibo, self.verbatim_circuit)`, letting `to_braket` handle all the translation. 

I hope this makes sense!

Reply:
Mmmh, weird, because the name of the gates should be standardized in QASM code. I mean, that's precisely the purpose of QASM: having a unified, framework agnostic, representation of a quantum circuit. In `qibo` we have a conversion layer from QASM to qibo gates and viceversa, and I assume AWS has it as well, therefore, assuming the QASM code we generate is compatible with the AWS implementation of QASM everything should be fine, up to minor corrections. For instance, if the `includes qe1lib.inc` import causes problems you could just drop it most likely.

Reply:
To follow up on the QASM discussion, I quickly tried running a random clifford circuit through the `braket` openqasm API and, except for the `No such file or directory: 'qelib1.inc'` error that can be overcome by simply dropping that line in the QASM string, everything seems to be working fine.

``` python
from braket.devices import LocalSimulator
device = LocalSimulator()
from braket.ir.openqasm import Program
from qibo.quantum_info import random_clifford
c = random_clifford(nqubits=3)
print(c.draw())
# q0: ─X─S───o───
# q1: ─H─Z───Z─o─
# q2: ─H─X─S───Z─
qasm_str = c.to_qasm()
qasm_str = "\n".join(qasm_str.split("\n")[3:])
program = Program(source=qasm_str)
my_task = device.run(program, shots=1000)
my_task.result().measurement_counts
# Counter({'111': 261, '101': 260, '110': 252, '100': 227})
```

Now this should be tested for other gates as well, such as the parametrized ones for instance, but I see no reason for them not to be working if `braket` is compatible with QASM as they say.

Reply:
Thanks @BrunoLiegiBastonLiegi for giving it a shot! 

I tried your code and it works for gates like H, X, S, and CZ, like in the example you commented out. However It doesn't work when there are CNOT gates and a few other gates like the sqrt(x) gate, which then require some translation of OpenQASM strings. 

For e.g., this circuit
``` python
q0: ─S─o─o─────o───
q1: ───|─|─H─S─|─o─
q2: ───Z─X─Y─S─Z─X─
``` 
does not work because it has a CNOT. The `ValueError: Gate cx is not defined.` will pop up.

Initially when I was using OpenQASM for translation, I searched manually for these gates which need explicit conversion and found these gates need translation:
``` python
_QASM_BRAKET_GATES = {
    "id": "i",
    "cx": "cnot",
    "sx": "v",
    "sxdg": "vi",
    "sdg": "si",
    "tdg": "ti",
    "u3": "U",
}
```

If I recall correctly, our colleague from AWS handled the Qiskit-Braket translation package. He did the Qiskit-to-Braket gate translation directly and he applied the same methodology for our Qibo-to-Braket translation package. Totally understand that OpenQASM should be the common ground for all languages, that's exactly my impression too. But sometimes each language seems to interpret the strings differently, so it may call for a different solution.

Hope to hear from you soon!

Reply:
Thanks, yep indeed I was able to verify myself that `CNOT` is causing troubles, weird choice from `braket` to use different names for the gates...
Anyway, a possible workaround to still try and rely on QASM is to manually correct for the problematic gates, for instance by doing:
``` python
import re
for k, v in _QASM_BRAKET_GATES.items():
    qasm_string = re.sub(f"\n{k} ", f"\n{v} ", qasm_string)
```
however this might be risky sometimes, what do you think @alecandido?

Reply:
To be fair: translating circuit is a pretty quick operation, and networking is a pretty expensive one.
Moreover, applying regexes or whatever other substitution to a possibly large chunk of text is definitely not that efficient anyhow... (you should implement a proper parser, but it would get more complex...)

So, I'd just go for:

Qibo -> OpenQASM -> Qiskit -> Braket

that should be pretty simple to implement, something like:
```py
qiskit_braket_provider.to_braket(QuantumCircuit.from_qasm_str(circuit.to_qasm()))
```

If then you'd like to implement a more direct translation for any specific reason, feel free to do that. But from my (limited) perspective, I'm not sure how much it will be used, so I'd save effort until it will be actually needed.
(notice that it would require an extra explicit dependency on Qiskit as well, to import the `QuantumCircuit` class)

Reply:
Hi @BrunoLiegiBastonLiegi ! 

> weird choice from braket to use different names for the gates...

I know, but the same can be said for Qiskit: Weird choice to use little endian. I suppose it is how each developer thinks it's the best way to implement for their company/language. 

Hi @alecandido ! Thanks for sharing your thoughts. 

> Qibo -> OpenQASM -> Qiskit -> Braket

Actually I would try to avoid having two extra dependencies in between Qibo and Braket. We saw that OpenQASM doesn't really work because of translation issues. And we'll need a proper parser like you mentioned for efficiency. I'm not sure if you've experienced it many people I know have commented that their codes were broken due to Qiskit's upgrading. So, although I don't think the `QuantumCircuit` class is affected much, I would prefer to avoid using Qiskit for now.

> If then you'd like to implement a more direct translation for any specific reason, feel free to do that. But from my (limited) perspective, I'm not sure how much it will be used, so I'd save effort until it will be actually needed.

Our colleague from AWS has done the direct translation Qibo -> Braket for us in `braket_translation.py`! We've tested it thoroughly and it works. Anyway, I believe @BrunoLiegiBastonLiegi and some of my IHPC colleagues intend to use the `braket_client` very soon. Do you think it be worth giving Qibo -> Braket a shot until further issues pop up?

Reply:
We had a chat about this, my take is the following (we should discuss in the next qibo meeting):
1. ask AWS to support qibo from their side.
2. use qiskit -> braket conversion (if we this allows control of transpilation)
3. go ahead with the current implementation.
We should prefer reducing maintenance.

Reply:
> We should prefer reducing maintenance.

I strongly agree with this statement.

Reply:
Stefano and I had a discussion just now and now I understand more clearly. Going from `qibo > qasm > qiskit > braket` has reduced maintenance. 

Here's the code to translate:
``` python
# Convert Qibo to OpenQASM
circuit_qasm = circuit_qibo.to_qasm()
# Convert OpenQASM to Qiskit
circuit_qiskit = QuantumCircuit.from_qasm_str(circuit_qasm)
# Convert Qiskit to Braket
circuit_braket = qiskit_braket_provider.to_braket(circuit_qiskit)
```

Putting them together, here's the code with verbatim circuits.
``` python
circuit_braket = qiskit_braket_provider.to_braket(QuantumCircuit.from_qasm_str(circuit_qibo.to_qasm()))

# For verbatim mode
if verbatim_circuit:

    # Extract Braket circuit without measurements
    circuit_braket_no_meas = BraketCircuit()
    for gate in circuit_braket.instructions:
        if gate.operator.name != "Measure":
            circuit_braket_no_meas.add_instruction(gate)
    
    # Add verbatim box
    circuit_braket_verbatim = BraketCircuit().add_verbatim_box(circuit_braket_no_meas)

    # Add the measurements
    for gate in circuit_braket.instructions:
        if gate.operator.name == "Measure":
            circuit_braket_verbatim.add_instruction(gate)
    
    circuit_braket = circuit_braket_verbatim
```

Only two gates I've found fail: `RYY`, `PRX`. The fault lies in the OpenQASM to Qiskit conversion.
`QASM2ParseError: "<input>:7,0: 'ryy' is not defined in this scope"`
`QASM2ParseError: "<input>:7,0: 'prx' is not defined in this scope"`

I tried to use QASM3 to import the qasm string to Qiskit circuit. However, `pip install qiskit_qasm3_import` breaks something in `qiskit_braket_provider` and `braket` with this error: `AttributeError: type object 'QASMNodeVisitor' has no attribute 'visitSwitchStatement'`

Reply:
> I tried to use QASM3 to import the qasm string to Qiskit circuit. However, `pip install qiskit_qasm3_import` breaks something in `qiskit_braket_provider` and `braket` with this error: `AttributeError: type object 'QASMNodeVisitor' has no attribute 'visitSwitchStatement'`

Please, limit as much as possible the direct dependencies. Qiskit should already be capable on its own to import a QASM circuit (the `QuantumCircuit.from_qasm_str()` above), and since QASM is definitely mainly sponsored by IBM, I'd say that it's fine to support whatever they are supporting, and retain all their limitation.

Moreover:
> <img width="588" alt="image" src="https://github.com/user-attachments/assets/f8f1ede3-6241-4027-a0a9-50dc3388ccdd">
>
> https://github.com/amazon-braket/amazon-braket-examples/blob/main/examples/braket_features/Verbatim_Compilation.ipynb

> When you run a quantum circuit on quantum computers from Rigetti, and IonQ, [...]
> https://docs.aws.amazon.com/braket/latest/developerguide/braket-openqasm-verbatim-compilation.html

So, it should not be relevant for IQM devices.

I'm also not sure why measurements have to be kept outside the verbatim block (what about mid-circuit measurements?).

And, eventually, since you may want to have just a verbatim section, not just a whole verbatim circuit, I'm not sure if the automated verbatim support really makes sense, since it's a bit more than the plain circuit execution, but much less than the full braket feature. So, maybe at that point (i.e. when you start requiring braket-specific instructions/directives) it would be much better for the user to directly dispatch a braket circuit, using their full API explicitly, instead of relying on an unclear translation by `qibo-cloud-backends`.

Reply:
> Please, limit as much as possible the direct dependencies. Qiskit should already be capable on its own to import a QASM circuit (the QuantumCircuit.from_qasm_str() above)

Yes indeed it works with limitations. I thought of trying using Qasm3 in case it allows translation of the PRX gates. But more issues were created with Qasm3. So let's just stick with `from_qasm_str` (Qasm2). 

> So, it should not be relevant for IQM devices.

I highly doubt this. The links might be outdated. Anyway, the lack of support for verbatim mode on IQM would have been highlighted by our colleagues from AWS. We ran some stuff with verbatim mode on IQM with no errors.

> why measurements have to be kept outside the verbatim block

Well, I tried to no avail. Just wrapping `braket_circuit` (including measurements) as follows
``` python
braket_circuit = BraketCircuit().add_verbatim_box(braket_circuit)
```
raises the `ValueError: cannot measure a subcircuit inside a verbatim box.` So it does look like we have to split up the circuit without measurements and measurements. In fact the same thing was done in `braket_translation`.

> mid-circuit measurements

Interesting suggestion! Perhaps let's discuss this extra feature along with other potentially useful features after the translation? I feel that it's a higher priority to resolve the translation. 

> you may want to have just a verbatim section, not just a whole verbatim circuit

Good point! I can give you a list of potential reasons why we want to have the entire circuit as verbatim, but let's discuss this after we resolve the translation issue.

Reply:
> I highly doubt this. The links might be outdated. Anyway, the lack of support for verbatim mode on IQM would have been highlighted by our colleagues from AWS. We ran some stuff with verbatim mode on IQM with no errors.

Well, if not unsupported, it is definitely undocumented (and thus private).
It seems rather unreliable to base a Qibo feature on such an ephemeral behavior...

If you really want to, you should directly accept braket circuits, but it would add pretty much nothing from the Qibo side...

> raises the `ValueError: cannot measure a subcircuit inside a verbatim box.` So it does look like we have to split up the circuit without measurements and measurements. In fact the same thing was done in `braket_translation`.

Notice that this would silently change the behavior of circuits containing mid-circuit measurements. If you really want to pursue this aim, you should validate the circuit to check you're not spoiling the user-specified instructions, and raise an error if they are unsupported.

> Interesting suggestion! Perhaps let's discuss this extra feature along with other potentially useful features after the translation? I feel that it's a higher priority to resolve the translation.

The single line I suggested above (passing ~7 comments before) should be enough to cover the translation, with no extra maintenance burden, and mostly supporting all required features. We're now discussing just extra features.

> Good point! I can give you a list of potential reasons why we want to have the entire circuit as verbatim, but let's discuss this after we resolve the translation issue.

Great, then lift this and all the other points above that are not going to be addressed in this PR as separate issues, and let's start to discuss them there :D

Reply:
Chiming in from the Braket team here:

1) The IQM device on Braket does support verbatim mode, and there's a section for it in the notebook you linked. That box you pointed to is a miss in the documentation and I've sent a PR to fix it.
2) If we do go the qiskit-braket-provider route, the `to_braket` function has a verbatim argument ([link](https://github.com/qiskit-community/qiskit-braket-provider/blob/903d83b6450295cce4f147d988333fb091583ebb/qiskit_braket_provider/providers/adapter.py#L454)), so we could use that to provide verbatim functionality. 
With this route then, the main issue is to get the circuit to Qiskit.
3) The Prx gate is Qiskit's R gate ([link](https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.library.RGate)). The existing QASM output name in Qibo may not match up with Qiskit so if we go Qibo > Qasm > Qiskit that may need to be fixed.

Minor comments (for informational purposes, feel free to skip):
3) Mid-circuit measurements are not supported on Braket at this point in time.
4) For the IonQ and IQM devices, the whole circuit has to be verbatim, except the final measurement instructions are outside the verbatim box

Reply:
Thanks for the update, I believe we know have enough information to make an informed choice.

> 1. The IQM device on Braket does support verbatim mode, and there's a section for it in the notebook you linked. That box Alessandro pointed to is a miss in the documentation and I've sent a PR to fix it.

Thanks also for this, it is useful to have an official confirmation (to avoid relying on hidden and possibly changing features).
In case, remember that it was still written in two independent places: the box in the notebook, and the paragraph in the web page (both links reported in https://github.com/qiboteam/qibo-cloud-backends/pull/30#discussion_r1760729625)

Reply:
Thanks @licedric ! 

> 2. If we do go the qiskit-braket-provider route, the to_braket function has a verbatim argument ([link](https://github.com/qiskit-community/qiskit-braket-provider/blob/903d83b6450295cce4f147d988333fb091583ebb/qiskit_braket_provider/providers/adapter.py#L454)), so we could use that to provide verbatim functionality.

Thanks for pointing out that qiskit-braket-provider has a verbatim function. I notice that verbatim box is added before the measurement gates.

> 3. The Prx gate is Qiskit's R gate ([link](https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.library.RGate)). The existing QASM output name in Qibo may not match up with Qiskit so if we go Qibo > Qasm > Qiskit that may need to be fixed.

Indeed, like what I commented earlier, only two gates I've found fail the OpenQASM to Qiskit conversion: RYY, PRX. The fault lies in the OpenQASM to Qiskit conversion.
`QASM2ParseError: "<input>:7,0: 'ryy' is not defined in this scope"`
`QASM2ParseError: "<input>:7,0: 'prx' is not defined in this scope"`

I will check and confirm soon.

Reply:
I agree that direct class-to-class translation is more robust than performing surgery on OpenQASM strings, but I'm also fine with using the Qiskit-Braket provider as the translation path; it just brings in an extra package, which feels like extra baggage.

Bringing in the provider also introduces a new path for transpilation; would it be preferable do that on the Qibo side or the Qiskit side?

Reply:
> I agree that direct class-to-class translation is more robust than performing surgery on OpenQASM strings, but I'm also fine with using the Qiskit-Braket provider as the translation path; it just brings in an extra package, which feels like extra baggage.

In general, I agree about avoiding extra deps. They are often a source of troubles in many updates.
However, this argument doesn't work forever: you would never avoid having NumPy as a dependency, and rewriting on your own even a few of their functions. And that's coming with a good motivation: reliable dependencies are avoiding code duplication and highly reducing the maintenance burden (deduplicating even that), that is the core of software reuse.

In this case, I'm not sure about how much person power the Qibo team will have to maintain this interface. So, unless someone else is taking over the burden, I'd consider Qiskit and the Qiskit-Braket interface as sufficiently reliable (i.e. a direct interface is certainly preferred, but not at the price of maintaining it, at the moment).

> Bringing in the provider also introduces a new path for transpilation; would it be preferable do that on the Qibo side or the Qiskit side?

Well, I know the long run answer: use Qibo. It is true that the backend will be something else, but this is a Qibo interface, and the UX should be as consistent as possible with Qibo.

However, our transpiler is still much work-in-progress, and that's a fact. Transitionally, it may be useful to adopt the Qiskit one. But this relies a lot on the planned use cases.

At the moment, it may even be difficult to impossible to use the Qibo transpiler for all possible platforms supported by Braket. I'd propose to move this item to a separate issue, and reconsider after solving the main one: plain execution.

Reply:
To follow up on `qiskit`'s `RYY` and `PRX/R` gates, I just quickly tried to check how their qasm representation looks like. Apparently, they don't have a single label assigned, but they are defined as new gates every time. A circuit containing an `RYY` and `R` gate is dumped to qasm like this:
```
OPENQASM 2.0;
include "qelib1.inc";
gate ryy(param0) q0,q1 { rx(pi/2) q0; rx(pi/2) q1; cx q0,q1; rz(param0) q1; cx q0,q1; rx(-pi/2) q0; rx(-pi/2) q1; }
gate r(param0,param1) q0 { u3(param0,param1 - pi/2,pi/2 - 1.0*param1) q0; }
qreg q[3];
ryy(0) q[1],q[2];
r(0,0) q[0];
``` 
which might make sense as these more exotic gates are more likely to not be directly supported in other frameworks, whereas their decomposition will probably be. However, this makes things significantly more annoying...

Reply:
> However, this makes things significantly more annoying...

Yes, we should prepare a preamble in `Circuit.to_qasm` and filter it according to the content of the `Circuit`.
But hard-coding the preamble would be redundant, while generating from some decomposition may be a bit involved...

Reply:
> which might make sense as these more exotic gates are more likely to not be directly supported in other frameworks, whereas their decomposition will probably be. However, this makes things significantly more annoying...

It seems to me that a better solution would be to skip the qasm step and go directly from Qibo to Qiskit; but that's probably outside the scope of this PR.

Reply:
For translating qasm to Qiskit, here is the `from_qasm_str` function that's currently in use: https://github.com/Qiskit/qiskit/blob/adf55e19193ec7798ada71fd17905229f0d0a6a8/qiskit/circuit/quantumcircuit.py#L4097

Note that it calls the `qasm2.loads` function, which is a more complete interface. You can define the set of gates to recognize through the `custom_instructions` argument; the default list is here: https://github.com/Qiskit/qiskit/blob/adf55e19193ec7798ada71fd17905229f0d0a6a8/qiskit/qasm2/parse.py#L153

If Qibo ever wants to use any gates not in that list, the `from_qasm_str` function will not work. I would suggest using the full `qasm2.loads` interface and specifying a list of custom instructions that is maintained by Qibo. For now this can simply be the `RYY` and `PRX/R` gates, but this list will surely be expanded in the future.

Reply:
Thanks once more: direct usage of `qasm2.loads()` could save us a lot of time with Qiskit-unsupported gates, making the translation a more viable and lower-cost alternative.

However, if Qiskit also supports the gates, but with different names, I would still consider the translation. Possibly even at the level of the Qibo QASM generator (if this is making our QASM more compatible, it may be worth).
Replacing `PRX` with `R` it could be way cheaper than any other alternative...

Reply:
> Replacing `PRX` with `R` it could be way cheaper than any other alternative...

I tried renaming output of Qibo's PRX as `r` but failed:

Error message is `QASM2ParseError: "<input>:7,0: 'r' is not defined in this scope"`

I suspect that `prx`/`r` is not listed inside the `LEGACY_CUSTOM_INSTRUCTIONS` here [https://github.com/Qiskit/qiskit/blob/adf55e19193ec7798ada71fd17905229f0d0a6a8/qiskit/qasm2/parse.py#L153](https://github.com/Qiskit/qiskit/blob/adf55e19193ec7798ada71fd17905229f0d0a6a8/qiskit/qasm2/parse.py#L153) which @licedric shared, causing the import `QuantumCircuit.from_qasm_str` to fail.

Reply:
Using this simple circuit,
``` python
from qiskit import qasm2
qc = QuantumCircuit(1)
qc.r(1, 2, 0)
qc_qasm = qasm2.dumps(qc)
print(qc_qasm)
```
the qasm string generated by qiskit is 
```
OPENQASM 2.0;
include "qelib1.inc";
gate r(param0,param1) q0 { u3(param0,param1 - pi/2,pi/2 - param1) q0; }
qreg q[1];
r(1.0,2.0) q[0];
```
We see that qiskit outputs `prx`/`r` gate as `r` in qasm.

Let's try reversing it. Importing `qc_qasm` to qiskit:
``` python
qc2 = QuantumCircuit.from_qasm_str(qc_qasm)
print(qc2)
```
Output:
``` python
   ┌────────┐
q: ┤ R(1,2) ├
   └────────┘
```
No error occurs.

Reply:
> At that point, it is maybe worth patching the `qibo` QASM generator, which would address both cases in one shot.

Yes, that's exactly what I had in mind.

> I was thinking about maybe adding a flag to our `circuit.to_qasm`, something like `decompose=True` which will trigger the decomposed definition of a pre-defined set of gates, `PRX` and `RYY` for now, the simple label would be used otherwise.

This is the slightly more painful way. For sure, we will need this for gates that are not supported by Qiskit, apparently (from @mho291 experiment above), we may need that also for the two incriminated gates.

But I wonder whether it's simpler to just copy the call inside `QuantumCircuit.from_qasm_str()`, and manually extend the `LEGACY_CUSTOM_INSTRUCTION`. I.e. something like

```py
    def from_qibo_qasm_str(qasm_str: str) -> "QuantumCircuit":
        """Convert a string containing an OpenQASM 2.0 program to a :class:`.QuantumCircuit`.
        """
        # pylint: disable=cyclic-import
        from qiskit import qasm2


        return qasm2.loads(
            qasm_str,
            include_path=qasm2.LEGACY_INCLUDE_PATH,
            custom_instructions=qasm2.LEGACY_CUSTOM_INSTRUCTIONS + our_custom_instructions,
            custom_classical=qasm2.LEGACY_CUSTOM_CLASSICAL,
            strict=False,
        )
```

Reply:
> But I wonder whether it's simpler to just copy the call inside QuantumCircuit.from_qasm_str(), and manually extend the LEGACY_CUSTOM_INSTRUCTION. I.e. something like

mmmh not sure to have got it right, but this would apply here only, right? For the sake of performing the `qibo -> qiskit -> braket` conversion. However, if you have, in general, a `qibo` circuit containing either `PRX` or `RYY`, you generate the QASM and try importing it to qiskit, it will fail.

Reply:
> But I wonder whether it's simpler to just copy the call inside `QuantumCircuit.from_qasm_str()`, and manually extend the `LEGACY_CUSTOM_INSTRUCTION`. I.e. something like

This is indeed what I had in mind. Maintain a list of Qibo custom instructions (currently at least PRX/R and RYY; really any gates not in `LEGACY_CUSTOM_INSTRUCTION`) and use Alessandro's code above to do the Qibo > Qasm > Qiskit translation, for both the Qiskit and Braket backends (so you could probably separate that out to its own function).

Reply:
For gates outside of `LEGACY_CUSTOM_INSTRUCTIONS` it doesn't actually matter what Qibo names it in Qasm. For example to get `prx` translated to Qiskit's R gate, I think we only need to define this in Qibo's custom instructions:
`CustomInstruction("prx", 2, 1, qiskit.circuit.library.RGate)`
which will define "prx" as a 2-parameter 1-qubit gate corresponding to Qiskit's `RGate` object.

Reply:
Exactly @licedric. My guess was that the `RGate`, with its Qiskit name, would actually be *inside* `LEGACY_CUSTOM_INSTRUCTIONS` (though I'm not that sure, since @mho291 showed that Qiskit itself is dumping it with the decomposition, though it is then representing it as a single gate after loading - so maybe the provided decomposition is ignored?).

So, the proposal would be to translate the names that have an equivalent in Qiskit, and extend `LEGACY_CUSTOM_INSTRUCTIONS` to complement all those gates that may be Qibo-specific (not just their names). If any.

@BrunoLiegiBastonLiegi the matter about what to dump directly with Qibo `to_qasm` is slightly less trivial.
If we want to use QASM as an interoperable format, we should forget about what is in Qiskit and `LEGACY_CUSTOM_INSTRUCTIONS` (since part of Qiskit), and just stick to the OpenQASM specification, providing the decompositions for those instructions that are not fully universal.
However, if the main reason why we care about QASM is just Qiskit compatibility, that's uselessly painful, and we can directly produce in Qibo some Qiskit-compatible QASM (though we'd make use of features beyond the specs).

Reply:
I am losing a bit the track here, so to summarize, the current proposal is:
- Performing the `qibo -> QASM -> qiskit -> braket` conversion
- To do that, implement a function, either inside this backend or outside but called inside the `execute_circuit` here
- The fucntion will build a `qiskit` circuit starting from the `qibo` generated QASM + some custom instructions through the `qasm2.loads` and overriding the `qiskit.from_qasm`  
- Finally, make use of the existing `qiskit -> braket` conversion layer to create the `braket` circuit and send it through

Did I miss anything?

Reply:
Thanks @BrunoLiegiBastonLiegi, yup that's right. We want to do `qibo -> qasm -> qiskit -> braket`. Here is what we know so far regarding the bottlenecks at `qasm -> qiskit`.

1. `qasm -> qiskit` does not work because the Phase RX gate is not in Qiskit's [`LEGACY_CUSTOM_INSTRUCTIONS`](https://github.com/Qiskit/qiskit/blob/adf55e19193ec7798ada71fd17905229f0d0a6a8/qiskit/qasm2/parse.py#L153). This means
    a. Qibo's PRX gate outputs `prx` as Qasm label fails to import to Qiskit
    b. Qibo's PRX gate outputs `r` as Qasm label fails to import to Qiskit

2. Regarding adding our `MY_CUSTOM_INSTRUCTIONS`, I tried this the following:
``` python
from qibo import Circuit as QiboCircuit, gates

circuit_qibo = QiboCircuit(2)
circuit_qibo.add(gates.H(0))
circuit_qibo.add(gates.CNOT(0,1))
circuit_qibo.add(gates.M(0, 1))
circuit_qibo.add(gates.PRX(0, 3, 4))

circuit_qasm = circuit_qibo.to_qasm()
print(circuit_qasm)
```
output:
```
// Generated by QIBO 0.2.12
OPENQASM 2.0;
include "qelib1.inc";
qreg q[2];
h q[0];
cx q[0],q[1];
r(3, 4) q[0];
```
Then,
``` python
from qiskit import qasm2
from qiskit.qasm2.parse import CustomInstruction
from qiskit.circuit import library as lib

MY_CUSTOM_INSTRUCTIONS = (CustomInstruction("r", 2, 1, lib.RGate),)

test_load = qasm2.loads(
    circuit_qasm,
    include_path=qasm2.LEGACY_INCLUDE_PATH,
    custom_instructions=qasm2.LEGACY_CUSTOM_INSTRUCTIONS + MY_CUSTOM_INSTRUCTIONS,
    custom_classical=qasm2.LEGACY_CUSTOM_CLASSICAL,
    strict=False,
)
```
Error pops up
```
QASM2ParseError: "<input>:7,0: cannot use non-builtin custom instruction 'r' before definition"
```